pkg/internal/cyberark/identity/identity.go

```diff
@@ -0,0 +1,441 @@
+
+	// Other fields omitted as they're not needed
+}
+
+// Client is an client for interacting with the CyberArk Identity API and performing a login using a username and password.
+// For context on the behaviour of this client, see the Pytho SDK: https://github.com/cyberark/ark-sdk-python/blob/3be12c3f2d3a2d0407025028943e584b6edc5996/ark_sdk_python/auth/identity/ark_identity.py
+type Client struct {
+	client *http.Client
+
+	endpoint  string
+	subdomain string
+
+	tokenCache      map[string]token
+	tokenCacheMutex sync.Mutex
+}
+
+// token is a wrapper type for holding auth tokens we want to cache.
+type token string
+
+// New returns an initialized CyberArk Identity client using a default service discovery client.
+// NB: This function performs service discovery when called, in order to ensure that all Identity
+// clients are created with a valid Identity API URL. This function blocks on the network call to
+// the discovery service.
+func New(ctx context.Context, subdomain string) (*Client, error) {
+	return NewWithDiscoveryClient(ctx, servicediscovery.New(), subdomain)
+}
+
+// NewWithDiscoveryClient returns an initialized CyberArk Identity client using the given service discovery client.
+// NB: This function performs service discovery when called, in order to ensure that all Identity
+// clients are created with a valid Identity API URL. This function blocks on the network call to
+// the discovery service.
+func NewWithDiscoveryClient(ctx context.Context, discoveryClient *servicediscovery.Client, subdomain string) (*Client, error) {
+	if discoveryClient == nil {
+		return nil, fmt.Errorf("must provide a non-nil discovery client to the Identity Client")
+	}
+
+	endpoint, err := discoveryClient.DiscoverIdentityAPIURL(ctx, subdomain)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Client{
+		client: &http.Client{
+			Timeout: 10 * time.Second,
+		},
+
+		endpoint:  endpoint,
+		subdomain: subdomain,
+
+		tokenCache:      make(map[string]token),
+		tokenCacheMutex: sync.Mutex{},
+	}, nil
+}
+
+// LoginUsernamePassword performs a blocking call to fetch an auth token from CyberArk Identity using the given username and password.
+// The password is zeroed after use.
+// Tokens are cached internally and are not directly accessible to code; use Client.AuthenticatedHTTPClient to add credentials
+// to an *http.Client.
+func (c *Client) LoginUsernamePassword(ctx context.Context, username string, password []byte) error {
+	defer func() {
+		for i := range password {
+			password[i] = 0x00
+		}
+	}()
+
+	c.tokenCacheMutex.Lock()
+	defer c.tokenCacheMutex.Unlock()
+
+	advanceRequestBody, err := c.doStartAuthentication(ctx, username)
+	if err != nil {
+		return err
+	}
+
+	// We can't skip AdvanceAuthentication so we need to add the password to the body
+	// and send it off to complete the login process
+	advanceRequestBody.Answer = string(password)
+
+	err = c.doAdvanceAuthentication(ctx, username, &advanceRequestBody)
```

Comment:
I think it would make more sense if the password bytes were supplied as an argument to this function.
Then the password bytes could be zeroed immediately after they are copied to the request body.